pkg/ebpf/btf.go

```diff
@@ -22,139 +22,68 @@ import (
-		Info:       0x0d000000,
-		SizeOrType: 0x0,
-		Extra: &pb.BtfType_Empty{
-			Empty: &pb.Empty{},
-		},
-	})
-
-	// 2: Func
-	types = append(types, &pb.BtfType{
-		NameOff:    0x1,
-		Info:       0x0c000000,
-		SizeOrType: 0x01,
-		Extra: &pb.BtfType_Empty{
-			Empty: &pb.Empty{},
-		},
-	})
-
-	// 3: Int
-	types = append(types, &pb.BtfType{
-		NameOff:    0x1,
-		Info:       0x01000000,
-		SizeOrType: 0x4,
-		Extra: &pb.BtfType_IntTypeData{
-			IntTypeData: &pb.IntTypeData{IntInfo: 0x01000020},
-		},
-	})
-
-	// 4: Struct
-	types = append(types, &pb.BtfType{
-		NameOff:    0x1,
-		Info:       0x04000001,
-		SizeOrType: 0x4,
-		Extra: &pb.BtfType_StructTypeData{
-			StructTypeData: &pb.StructTypeData{
-				NameOff:    0x1,
-				StructType: 0x3,
-				Offset:     0x0,
-			},
-		},
-	})
-
-	// 5: Ptr
-	types = append(types, &pb.BtfType{
-		NameOff:    0x0,
-		Info:       0x02000000,
-		SizeOrType: 0x4,
-		Extra: &pb.BtfType_Empty{
-			Empty: &pb.Empty{},
-		},
-	})
-
-	// 6: Func_Proto
-	types = append(types, &pb.BtfType{
-		NameOff:    0x0,
-		Info:       0x0d000002,
-		SizeOrType: 0x3,
-		Extra: &pb.BtfType_FuncProtoTypeData{
-			FuncProtoTypeData: &pb.FuncProtoTypeData{
-				Param: []*pb.BtfParam{
-					{NameOff: 0x1, ParamType: 0x3},
-					{NameOff: 0x1, ParamType: 0x5},
-				},
-			},
-		},
-	})
+// SetHeaderSection takes a BTF proto and assigns the header values, magic
+// represents if the system is big or little endian (0xeb9f, 0x9feb).
+func SetHeaderSection(btf *pb.Btf, magic int32, version int32, flags int32) {
+	btf.Header = &pb.Header{
+		Magic:   magic,
+		Version: version,
+		Flags:   flags,
+		HdrLen:  0x18,
+		TypeOff: 0x0,
+		TypeLen: int32(proto.Size(btf.TypeSection)),
+		StrOff:  int32(proto.Size(btf.TypeSection)),
+		StrLen:  int32(proto.Size(btf.StringSection)),
+	}
+}
 
-	// 7: Func
-	types = append(types, &pb.BtfType{
-		NameOff:    0x1,
-		Info:       0x0c000000,
-		SizeOrType: 0x6,
-		Extra: &pb.BtfType_Empty{
-			Empty: &pb.Empty{},
-		},
-	})
-	return &pb.TypeSection{BtfType: types}
+// SetTypeSection populates the BTF type section of the provided BTF proto with
+// the parameter types
+func SetTypeSection(btf *pb.Btf, types []*pb.BtfType) {
+	btf.TypeSection = &pb.TypeSection{
+		BtfType: types,
+	}
 }
 
-func stringSection() *pb.StringSection {
-	// For buzzer's BTF Implementation the string sections is not currently
-	// in used.
-	return &pb.StringSection{Str: "buzzer"}
+// SetTypeInfo returns the parameter type info encoded,
+// the bits arrangement is:
+//   - bits 0-15  = vlen
+//   - bits 16-23 = unused
+//   - bits 24-28 = kind
+//   - bits 29-30 = unused
+//   - bits    31 = kind_flag
+//
+// https://docs.kernel.org/bpf/btf.html#:~:text=Each%20type%20contains%20the%20following%20common%20data%3A
+func SetTypeInfo(vlen int16, kind pb.BtfKind, kind_flag bool) int32 {
+	info := int32(0)
+	info |= int32(vlen)
+	info |= int32(kind) << 24
+	flag := 0
+	if kind_flag {
+		flag = 1
+	}
+	info |= int32(flag) << 28
```

Comment:
we could simplify this even further by putting it inside the if statement (instead of declaring a flag local variable)

Reply:
Done, changed it into a if/else